File: queen-attack/queen_attack.rb
class Queens

  DEFAULT_OPTIONS = {
    white: [0,3],
    black: [7,3]
  }.freeze

  def initialize(options = {})
    @options = DEFAULT_OPTIONS.merge(options)

    raise ArgumentError, "Pieces can't occupy the same space" if same_space?
  end

  def white
    @options[:white]
  end

  def black
    @options[:black]
  end

  def attack?
    same_row? || same_column? || same_diagonal?
  end

  def to_s
    (0..7).map do |row|
      (0..7).map do |col|
        case
        when white == [row,col] then "W"
        when black == [row,col] then "B"
        else "_"
        end
      end.join(" ")
    end.join("\n")
  end

  private

  def same_space?
    white == black
  end

  def same_row?
    white[0] == black[0]
  end

  def same_column?
    white[1] == black[1]
  end

  def same_diagonal?
    (black[0] - white[0]).abs == (black[1] - white[1]).abs
  end

end

File: elixir/queen-attack/queen_attack.exs
defmodule Queens do
  @type t :: %Queens{ black: {integer, integer}, white: {integer, integer} }
  defstruct black: nil, white: nil

  @doc """
  Creates a new set of Queens
  """
  @spec new() :: Queens.t()
  @spec new({integer, integer}, {integer, integer}) :: Queens.t()
  def new(white \\ {0,3}, black \\ {7,3})
  def new(white, black) when white == black, do: raise ArgumentError
  def new(white, black) do
    %Queens{ white: white, black: black }
  end

  @doc """
  Gives a string reprentation of the board with
  white and black queen locations shown
  """
  @spec to_string(Queens.t()) :: String.t()
  def to_string(queens) do
    Enum.map_join(0..7, "\n", &row(&1, queens))
  end

  defp row(row_num, queens), do: Enum.map_join(0..7, " ", &cell(row_num, &1, queens))

  defp cell(row, col, %{black: {row,col}}), do: "B"
  defp cell(row, col, %{white: {row,col}}), do: "W"
  defp cell(_row, _col, _queens), do: "_"

  @doc """
  Checks if the queens can attack each other
  """
  @spec can_attack?(Queens.t()) :: boolean
  def can_attack?(%{black: {row, _}, white: {row, _}}), do: true
  def can_attack?(%{black: {_, col}, white: {_, col}}), do: true
  def can_attack?(%{black: {black_row, black_col}, white: {white_row, white_col}}) do
    abs(black_row - white_row) == abs(black_col - white_col)
  end
end
